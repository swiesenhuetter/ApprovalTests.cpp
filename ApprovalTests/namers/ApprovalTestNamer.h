#pragma once

#include "ApprovalTests/core/ApprovalNamer.h"
#include <sstream>
#include <vector>
#include <stdexcept>
#include "ApprovalTests/utilities/SystemUtils.h"
#include "ApprovalTests/namers/HelpMessages.h"

namespace ApprovalTests
{
    class TestName
    {
    public:
        const std::string& getFileName() const
        {
            checkBuildConfiguration(fileName);
            return fileName;
        }

        void setFileName(const std::string& file)
        {
            fileName = SystemUtils::checkFilenameCase(file);
        }

    private:
        static void checkBuildConfiguration(const std::string& fileName)
        {
            if (!FileUtils::fileExists(fileName))
            {
                throw std::runtime_error(getMisconfiguredBuildHelp(fileName));
            }
        }

    public:
        static std::string getMisconfiguredBuildHelp(const std::string& fileName)
        {
            return "\n\n" + HelpMessages::getMisconfiguredBuildHelp(fileName) + "\n\n";
        }

        std::vector<std::string> sections;

    private:
        std::string fileName;
    };

    class TestConfiguration
    {
    public:
        std::string subdirectory;
    };

    class ApprovalTestNamer : public ApprovalNamer
    {
    private:
    public:
        ApprovalTestNamer() = default;

        std::string getTestName() const
        {
            std::stringstream ext;
            auto test = getCurrentTest();
            for (size_t i = 0; i < test.sections.size(); i++)
            {
                if (0 < i)
                {
                    ext << ".";
                }
                ext << test.sections[i];
            }

            return convertToFileName(ext.str());
        }

        static bool isForbidden(char c)
        {
            static std::string forbiddenChars("\\/:?\"<>|' ");
            return std::string::npos != forbiddenChars.find(c);
        }

        static std::string convertToFileName(const std::string& fileName)
        {
            std::stringstream result;
            for (auto ch : fileName)
            {
                if (!isForbidden(ch))
                {
                    result << ch;
                }
                else
                {
                    result << "_";
                }
            }
            return result.str();
        }

        static TestName& getCurrentTest()
        {
            try
            {
                return currentTest();
            }
            catch (const std::runtime_error&)
            {
                std::string helpMessage = getMisconfiguredMainHelp();
                throw std::runtime_error(helpMessage);
            }
        }

        static std::string getMisconfiguredMainHelp()
        {
            return "\n\n" + HelpMessages::getMisconfiguredMainHelp() + "\n\n";
        }

        // Deprecated - please use getSourceFileName
        std::string getFileName() const
        {
            return getSourceFileName();
        }

        std::string getSourceFileName() const
        {
            auto file = getCurrentTest().getFileName();
            auto start = file.rfind(SystemUtils::getDirectorySeparator()) + 1;
            auto end = file.rfind('.');
            auto fileName = file.substr(start, end - start);
            return convertToFileName(fileName);
        }

        std::string getDirectory() const
        {
            auto file = getCurrentTest().getFileName();
            auto end = file.rfind(SystemUtils::getDirectorySeparator()) + 1;
            auto directory = file.substr(0, end);
            if (!testConfiguration().subdirectory.empty())
            {
                directory += testConfiguration().subdirectory +
                             SystemUtils::getDirectorySeparator();
                SystemUtils::ensureDirectoryExists(directory);
            }
            return directory;
        }

        static TestName& currentTest(TestName* value = nullptr)
        {
            static TestName* staticValue;
            if (value != nullptr)
            {
                staticValue = value;
            }
            if (staticValue == nullptr)
            {
                throw std::runtime_error(
                    "The variable in currentTest() is not initialised");
            }
            return *staticValue;
        }

        static TestConfiguration& testConfiguration()
        {
            static TestConfiguration configuration;
            return configuration;
        }

        virtual std::string getApprovedFile(std::string extensionWithDot) const override
        {

            return getFullFileName(".approved", extensionWithDot);
        }

        virtual std::string getReceivedFile(std::string extensionWithDot) const override
        {

            return getFullFileName(".received", extensionWithDot);
        }

        std::string getOutputFileBaseName() const
        {
            return getSourceFileName() + "." + getTestName();
        }

        std::string getFullFileName(const std::string& approved,
                                    const std::string& extensionWithDot) const
        {
            std::stringstream ext;
            ext << getDirectory() << getOutputFileBaseName() << approved
                << extensionWithDot;
            return ext.str();
        }
    };
}
