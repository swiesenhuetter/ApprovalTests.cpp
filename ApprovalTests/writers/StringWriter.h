#pragma once

#include <string>
#include <fstream>
#include <stdexcept>
#include <utility>
#include "ApprovalTests/core/ApprovalWriter.h"

namespace ApprovalTests
{
    class StringWriter : public ApprovalWriter
    {
    private:
        std::string s;
        std::string ext;

    public:
        explicit StringWriter(std::string contents,
                              std::string fileExtensionWithDot = ".txt")
            : s(std::move(contents)), ext(std::move(fileExtensionWithDot))
        {
        }

        std::string getFileExtensionWithDot() const override
        {
            return ext;
        }

        void write(std::string path) const override
        {
            std::ofstream out(path.c_str(), std::ofstream::out);
            if (!out)
            {
                throw std::runtime_error("Unable to write file: " + path);
            }
            this->Write(out);
            out.close();
        }

        void Write(std::ostream& out) const
        {
            out << s << "\n";
        }

        virtual void cleanUpReceived(std::string receivedPath) const override
        {
            remove(receivedPath.c_str());
        }
    };
}
