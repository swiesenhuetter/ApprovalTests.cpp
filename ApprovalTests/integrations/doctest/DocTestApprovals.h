#pragma once

#include "ApprovalTests/namers/ApprovalTestNamer.h"
#include "ApprovalTests/integrations/CheckFileMacroIsAbsolute.h"

#if defined(APPROVALS_DOCTEST_EXISTING_MAIN)
#define APPROVALS_DOCTEST
#define DOCTEST_CONFIG_IMPLEMENT
#elif defined(APPROVALS_DOCTEST)
#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
#endif

#ifdef APPROVALS_DOCTEST

// begin-snippet: required_header_for_doctest
#include <doctest/doctest.h>
// end-snippet

namespace ApprovalTests
{
    // anonymous namespace to prevent compiler -Wsubobject-linkage warnings
    // This is OK as this code is only compiled on main()
    namespace
    {
        struct AbstractReporter : doctest::IReporter
        {
            virtual void report_query(const doctest::QueryData&) override
            {
            }
            // called when the whole test run starts
            virtual void test_run_start() override
            {
            }

            // called when the whole test run ends (caching a pointer to the input doesn't make sense here)
            virtual void test_run_end(const doctest::TestRunStats&) override
            {
            }

            // called when a test case is started (safe to cache a pointer to the input)
            virtual void test_case_start(const doctest::TestCaseData&) override
            {
            }

#if 20305 <= DOCTEST_VERSION
            // called when a test case is reentered because of unfinished subcases (safe to cache a pointer to the input)
            virtual void test_case_reenter(const doctest::TestCaseData&) override
            {
            }
#endif

            // called when a test case has ended
            virtual void test_case_end(const doctest::CurrentTestCaseStats&) override
            {
            }

            // called when an exception is thrown from the test case (or it crashes)
            virtual void test_case_exception(const doctest::TestCaseException&) override
            {
            }

            // called whenever a subcase is entered (don't cache pointers to the input)
            virtual void subcase_start(const doctest::SubcaseSignature&) override
            {
            }

            // called whenever a subcase is exited (don't cache pointers to the input)
            virtual void subcase_end() override
            {
            }

            // called for each assert (don't cache pointers to the input)
            virtual void log_assert(const doctest::AssertData&) override
            {
            }

            // called for each message (don't cache pointers to the input)
            virtual void log_message(const doctest::MessageData&) override
            {
            }

            // called when a test case is skipped either because it doesn't pass the filters, has a skip decorator
            // or isn't in the execution range (between first and last) (safe to cache a pointer to the input)
            virtual void test_case_skipped(const doctest::TestCaseData&) override
            {
            }
        };

        struct DocTestApprovalListener : AbstractReporter
        {
            TestName currentTest;

            // constructor has to accept the ContextOptions by ref as a single argument
            explicit DocTestApprovalListener(const doctest::ContextOptions& /*in*/)
            {
            }

            void test_case_start(const doctest::TestCaseData& testInfo) override
            {

                currentTest.sections.emplace_back(testInfo.m_name);
                currentTest.setFileName(testInfo.m_file);
                ApprovalTestNamer::currentTest(&currentTest);
            }

            void test_case_end(const doctest::CurrentTestCaseStats& /*in*/) override
            {

                while (!currentTest.sections.empty())
                {
                    currentTest.sections.pop_back();
                }
            }

            void subcase_start(const doctest::SubcaseSignature& signature) override
            {

#if DOCTEST_VERSION >= 20307
                currentTest.sections.emplace_back(signature.m_name.c_str());
#else
                currentTest.sections.emplace_back(signature.m_name);
#endif
            }

            void subcase_end() override
            {

                currentTest.sections.pop_back();
            }
        };
    }
}

REGISTER_LISTENER("approvals", 0, ApprovalTests::DocTestApprovalListener);

#endif // APPROVALS_DOCTEST
