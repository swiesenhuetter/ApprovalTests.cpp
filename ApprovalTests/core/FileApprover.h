#pragma once

#include <memory>
#include "ApprovalException.h"
#include "ApprovalTests/writers/StringWriter.h"
#include "Reporter.h"
#include "ApprovalTests/reporters/FrontLoadedReporterDisposer.h"
#include "ApprovalTests/core/ApprovalNamer.h"
#include "ApprovalComparator.h"
#include "ApprovalTests/comparators/ComparatorFactory.h"
#include "ApprovalTests/utilities/FileUtils.h"

namespace ApprovalTests
{

    class FileApprover
    {

    public:
        FileApprover() = default;

        ~FileApprover() = default;

        static ComparatorDisposer
        registerComparatorForExtension(const std::string& extensionWithDot,
                                       std::shared_ptr<ApprovalComparator> comparator)
        {
            return ComparatorFactory::registerComparator(extensionWithDot, comparator);
        }

        //! This overload is an implementation detail. To add a new comparator, use registerComparator().
        static void verify(const std::string& receivedPath,
                           const std::string& approvedPath,
                           const ApprovalComparator& comparator)
        {
            if (!FileUtils::fileExists(approvedPath))
            {
                throw ApprovalMissingException(receivedPath, approvedPath);
            }

            if (!FileUtils::fileExists(receivedPath))
            {
                throw ApprovalMissingException(approvedPath, receivedPath);
            }

            if (!comparator.contentsAreEquivalent(receivedPath, approvedPath))
            {
                throw ApprovalMismatchException(receivedPath, approvedPath);
            }
        }

        static void verify(const std::string& receivedPath,
                           const std::string& approvedPath)
        {
            verify(receivedPath,
                   approvedPath,
                   *ComparatorFactory::getComparatorForFile(receivedPath));
        }

        static void
        verify(const ApprovalNamer& n, const ApprovalWriter& s, const Reporter& r)
        {
            std::string approvedPath = n.getApprovedFile(s.getFileExtensionWithDot());
            std::string receivedPath = n.getReceivedFile(s.getFileExtensionWithDot());
            s.write(receivedPath);
            try
            {
                verify(receivedPath, approvedPath);
                s.cleanUpReceived(receivedPath);
            }
            catch (const ApprovalException&)
            {
                reportAfterTryingFrontLoadedReporter(receivedPath, approvedPath, r);
                throw;
            }
        }

        static void reportAfterTryingFrontLoadedReporter(const std::string& receivedPath,
                                                         const std::string& approvedPath,
                                                         const Reporter& r)
        {
            auto tryFirst = FrontLoadedReporterFactory::getFrontLoadedReporter();
            if (!tryFirst->report(receivedPath, approvedPath))
            {
                r.report(receivedPath, approvedPath);
            }
        }
    };
}
