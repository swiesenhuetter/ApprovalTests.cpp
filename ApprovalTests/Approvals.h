#pragma once

#include <string>
#include <functional>
#include <exception>

#include "ApprovalTests/ApprovalTestsVersion.h"
#include "ApprovalTests/core/FileApprover.h"
#include "reporters/DefaultReporter.h"
#include "reporters/DefaultReporterDisposer.h"
#include "ApprovalTests/core/Reporter.h"
#include "namers/ApprovalTestNamer.h"
#include "writers/ExistingFile.h"
#include "namers/ExistingFileNamer.h"
#include "namers/SubdirectoryDisposer.h"
#include "namers/DefaultNamerDisposer.h"
#include "scrubbers/Scrubbers.h"

namespace ApprovalTests
{

    class Approvals
    {
    private:
        Approvals() = default;

        ~Approvals() = default;

    public:
        static std::shared_ptr<ApprovalNamer> getDefaultNamer()
        {
            return DefaultNamerFactory::getDefaultNamer()();
        }

        static void verify(std::string contents,
                           const Reporter& reporter = DefaultReporter())
        {
            verifyScrubbed(contents, Scrubbers::doNothing, reporter);
        }

        static void verifyScrubbed(std::string contents,
                                   Scrubber scrubber,
                                   const Reporter& reporter = DefaultReporter())
        {
            verifyWithExtension(scrubber(contents), ".txt", reporter);
        }

        static void verifyWithExtension(std::string contents,
                                        const std::string& fileExtensionWithDot,
                                        const Reporter& reporter = DefaultReporter())
        {
            StringWriter writer(contents, fileExtensionWithDot);
            FileApprover::verify(*getDefaultNamer(), writer, reporter);
        }

        static void verify(const ApprovalWriter& writer,
                           const Reporter& reporter = DefaultReporter())
        {
            FileApprover::verify(*getDefaultNamer(), writer, reporter);
        }

        template <typename T>
        using IsNotDerivedFromWriter =
            typename std::enable_if<!std::is_base_of<ApprovalWriter, T>::value,
                                    int>::type;

        template <typename T, typename = IsNotDerivedFromWriter<T>>
        static void verify(const T& contents,
                           const Reporter& reporter = DefaultReporter())
        {
            verify(StringUtils::toString(contents), reporter);
        }

        template <typename T, typename = IsNotDerivedFromWriter<T>>
        static void verifyWithExtension(const T& contents,
                                        const std::string& fileExtensionWithDot,
                                        const Reporter& reporter = DefaultReporter())
        {
            verifyWithExtension(
                StringUtils::toString(contents), fileExtensionWithDot, reporter);
        }

        template <typename T,
                  typename Function,
                  typename = Detail::EnableIfNotDerivedFromReporter<Function>>
        static void verify(const T& contents,
                           Function converter,
                           const Reporter& reporter = DefaultReporter())
        {
            std::stringstream s;
            converter(contents, s);
            verify(s.str(), reporter);
        }

        template <typename T,
                  typename Function,
                  typename = Detail::EnableIfNotDerivedFromReporter<Function>>
        static void verifyWithExtension(const T& contents,
                                        Function converter,
                                        const std::string& fileExtensionWithDot,
                                        const Reporter& reporter = DefaultReporter())
        {
            std::stringstream s;
            converter(contents, s);
            verifyWithExtension(s.str(), fileExtensionWithDot, reporter);
        }

        static void verifyExceptionMessage(std::function<void(void)> functionThatThrows,
                                           const Reporter& reporter = DefaultReporter())
        {
            std::string message = "*** no exception thrown ***";
            try
            {
                functionThatThrows();
            }
            catch (const std::exception& e)
            {
                message = e.what();
            }
            verify(message, reporter);
        }

        template <typename Iterator>
        static void verifyAll(
            std::string header,
            const Iterator& start,
            const Iterator& finish,
            std::function<void(typename Iterator::value_type, std::ostream&)> converter,
            const Reporter& reporter = DefaultReporter())
        {
            std::stringstream s;
            if (!header.empty())
            {
                s << header << "\n\n\n";
            }
            for (auto it = start; it != finish; ++it)
            {
                converter(*it, s);
                s << '\n';
            }
            verify(s.str(), reporter);
        }

        template <typename Container>
        static void verifyAll(
            std::string header,
            const Container& list,
            std::function<void(typename Container::value_type, std::ostream&)> converter,
            const Reporter& reporter = DefaultReporter())
        {
            verifyAll<typename Container::const_iterator>(
                header, list.begin(), list.end(), converter, reporter);
        }

        template <typename T>
        static void verifyAll(std::string header,
                              const std::vector<T>& list,
                              const Reporter& reporter = DefaultReporter())
        {
            int i = 0;
            verifyAll<std::vector<T>>(
                header,
                list,
                [&](T e, std::ostream& s) { s << "[" << i++ << "] = " << e; },
                reporter);
        }

        template <typename T>
        static void verifyAll(const std::vector<T>& list,
                              const Reporter& reporter = DefaultReporter())
        {
            verifyAll<T>("", list, reporter);
        }

        static void verifyExistingFile(const std::string filePath,
                                       const Reporter& reporter = DefaultReporter())
        {
            ExistingFile writer(filePath);
            ExistingFileNamer namer(filePath);
            FileApprover::verify(namer, writer, reporter);
        }

        static SubdirectoryDisposer
        useApprovalsSubdirectory(std::string subdirectory = "approval_tests")
        {
            return SubdirectoryDisposer(subdirectory);
        }

        static DefaultReporterDisposer
        useAsDefaultReporter(const std::shared_ptr<Reporter>& reporter)
        {
            return DefaultReporterDisposer(reporter);
        }

        static FrontLoadedReporterDisposer
        useAsFrontLoadedReporter(const std::shared_ptr<Reporter>& reporter)
        {
            return FrontLoadedReporterDisposer(reporter);
        }

        static DefaultNamerDisposer useAsDefaultNamer(NamerCreator namerCreator)
        {
            return DefaultNamerDisposer(namerCreator);
        }
    };
}
