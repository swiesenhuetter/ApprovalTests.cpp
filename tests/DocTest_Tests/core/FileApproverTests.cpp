#include "doctest/doctest.h"
#include "ApprovalTests/writers/StringWriter.h"
#include "../../DocTest_Tests/reporters/TestReporter.h"
#include "ApprovalTests/namers/ApprovalTestNamer.h"
#include "ApprovalTests/core/FileApprover.h"
#include "ApprovalTests/Approvals.h"

using namespace ApprovalTests;

TEST_CASE("ItVerifiesApprovedFileExists")
{

    ApprovalTestNamer namer;
    StringWriter writer("Hello");
    TestReporter reporter;

    std::string approved = namer.getApprovedFile(".txt");
    std::string received = namer.getReceivedFile(".txt");

    std::string expected = "Failed Approval: \n"
                           "Approval File Not Found \n"
                           "File: \"" +
                           approved + "\"";
    REQUIRE_THROWS_MESSAGE(FileApprover::verify(namer, writer, reporter), expected);

    remove(approved.c_str());
    remove(received.c_str());
}

TEST_CASE("ItVerifiesExistingFiles")
{
    ApprovalTestNamer namer;
    Approvals::verifyExistingFile(namer.getDirectory() + "../../sample.txt");
}

TEST_CASE("ItIgnoresLineEndingDifferences")
{
    FileUtils::writeToFile("a.txt", "1\r\n2\n3\r\n4\r\n5");
    FileUtils::writeToFile("b.txt", "1\n2\r\n3\r\n4\n5");
    FileApprover::verify("a.txt", "b.txt");
}

TEST_CASE("ItComparesTheEntireFile")
{
    FileUtils::writeToFile("a.txt", "12345");
    FileUtils::writeToFile("b.txt", "123");
    CHECK_THROWS_AS(FileApprover::verify("a.txt", "b.txt"), ApprovalMismatchException);
}

// begin-snippet: create_custom_comparator
class LengthComparator : public ApprovalComparator
{
public:
    bool contentsAreEquivalent(std::string receivedPath,
                               std::string approvedPath) const override
    {
        return FileUtils::fileSize(receivedPath) == FileUtils::fileSize(approvedPath);
    }
};
// end-snippet

TEST_CASE("ItUsesCustomComparator")
{
    FileUtils::writeToFile("a.length", "12345");
    FileUtils::writeToFile("b.length", "56789");

    // begin-snippet: use_custom_comparator
    auto disposer = FileApprover::registerComparatorForExtension(
        ".length", std::make_shared<LengthComparator>());
    // end-snippet

    FileApprover::verify("a.length", "b.length");
}
