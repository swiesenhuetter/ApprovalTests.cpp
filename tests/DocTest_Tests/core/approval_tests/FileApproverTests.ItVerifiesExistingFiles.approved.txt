Everytime something repeatitive isn't automated, a manager gets hired.
